evalml/automl/automl_algorithm/default_algorithm.py

```diff
@@ -408,6 +408,43 @@ def next_batch(self):
         self._batch_number += 1
         return next_batch
 
+    def _get_feature_provenance_and_remove_engineered_features(
+        self, pipeline, component_name, to_be_removed, to_be_added
+    ):
+        component = pipeline.get_component(component_name)
+        feature_provenance = component._get_feature_provenance()
+        for original_col in feature_provenance:
+            selected = False
+            for encoded_col in feature_provenance[original_col]:
+                if encoded_col in to_be_removed:
+                    selected = True
+                    to_be_removed.remove(encoded_col)
+            if selected:
+                to_be_added.append(original_col)
+
+    def _parse_selected_categorical_features(self, pipeline):
+        if list(self.X.ww.select("categorical").columns):
+            self._get_feature_provenance_and_remove_engineered_features(
```

Comment:
I think this is good but I feel like we need something more general here long term. If a new logical type is introduced, we might fall into the same bug again. Let's file an ice box issue to not forget.  I can try to elaborate more there.